flocker/node/agents/gce.py

```diff
@@ -0,0 +1,385 @@
+        """
+        Initialize the GCEBlockDeviceAPI.
+
+        :param unicode project: The project where all GCE operations will take
+            place.
+        :param unicode zone: The zone where all GCE operations will take place.
+        """
+        # TODO(mewert): Also enable credentials via service account private
+        # keys.
+        credentials = AppAssertionCredentials(
+            "https://www.googleapis.com/auth/cloud-platform")
+        self._compute = discovery.build(
+            'compute', 'v1', credentials=credentials)
+        self._project = project
+        self._zone = zone
+        self._cluster_id = cluster_id
+
+    def _disk_resource_description(self):
+        """
+        Returns the value to be used in the description field of the disk
+        resources for this cluster.
+
+        :returns unicode: The value for the description.
+        """
+        return u"flocker-v1-cluster-id: " + unicode(self._cluster_id)
+
+    def _do_blocking_operation(self, function, **kwargs):
+        """
+        Perform a GCE operation, blocking until the operation completes.
+
+        This will call `function` with the passed in keyword arguments plus
+        additional keyword arguments for project and zone which come from the
+        private member variables with the same name. It is expected that
+        `function` returns an object that has an `execute()` method that
+        returns a GCE operation resource dict.
+
+        This function will then poll the operation until it reaches state
+        'DONE' or times out, and then returns the final operation resource
+        dict.
+
+        :param function: Callable that takes keyword arguments project and
+            zone, and returns an executable that results in a GCE operation
+            resource dict as described above.
+        :param kwargs: Additional keyword arguments to pass to function.
+
+        :returns dict: A dict representing the concluded GCE operation
+            resource.
+        """
+        # TODO(mewert): Be more sophisticated about timeout and retry loop.
+        # Look at EBS code, read up on how GCE behaves, potentially allow each
+        # operation to specify its own timeout. Also pass a reactor in so you
+        # can test the timeout error paths in unit tests. Also document what
+        # happens on timeout.
+        args = dict(project=self._project, zone=self._zone)
+        args.update(kwargs)
+        operation = function(**args).execute()
+        operation_name = operation['name']
+
+        def finished_operation_result():
+            latest_operation = self._compute.zoneOperations().get(
+                project=self._project,
+                zone=self._zone,
+                operation=operation_name).execute()
+            # TODO Logging
+            if latest_operation['status'] == 'DONE':
+                return latest_operation
+            return None
+
+        return poll_until(finished_operation_result, [1]*35)
```

Comment:
We plan on gathering sampling latency for various operations inside GCE and use that to create better informed timeouts.